File: android/java/com/tcc/api/LarkClient.java
package com.tcc.api;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

/**
 * Lightweight client for executing lark-cli commands.
 * lark-cli must be installed on the system (e.g., in Termux).
 */
public class LarkClient {

    private static final String LARK_CLI = "lark-cli";

    /**
     * Check if lark-cli is available on the system.
     *
     * @return true if lark-cli can be executed
     */
    public static boolean isAvailable() {
        try {
            Process process = new ProcessBuilder("which", LARK_CLI)
                    .redirectErrorStream(true)
                    .start();
            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
            String line = reader.readLine();
            reader.close();
            process.waitFor();
            return line != null && !line.isEmpty();
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Execute a lark-cli command and return stdout as a string.
     *
     * @param args The command arguments (e.g., {"auth", "status"})
     * @return The stdout output as a string
     */
    public static String execute(String[] args) {
        if (args == null) {
            return "";
        }

        String[] cmd = new String[args.length + 1];
        cmd[0] = LARK_CLI;
        System.arraycopy(args, 0, cmd, 1, args.length);

        try {
            Process process = new ProcessBuilder(cmd)
                    .redirectErrorStream(true)
                    .start();

            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
            StringBuilder output = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                if (output.length() > 0) {
                    output.append('\n');
                }
                output.append(line);
            }
            reader.close();

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                return output.toString();
            }

            return output.toString();

        } catch (IOException e) {
            return "Error: " + e.getMessage();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return "Error: Process interrupted";
        }
    }

    /**
     * Execute a lark-cli command and parse the output as JSON.
     *
     * @param args The command arguments
     * @return A JSONObject parsed from stdout, or an empty JSONObject on failure
     */
    public static JSONObject executeJson(String[] args) {
        String output = execute(args);
        if (output == null || output.isEmpty()) {
            return new JSONObject();
        }
        try {
            return new JSONObject(output);
        } catch (JSONException e) {
            JSONObject error = new JSONObject();
            try {
                error.put("error", "Failed to parse output as JSON");
                error.put("raw", output);
            } catch (JSONException je) {
                // Should not happen
            }
            return error;
        }
    }

    /**
     * Execute a lark-cli command with a specific working directory.
     *
     * @param args The command arguments
     * @param workingDir The working directory for the process
     * @return The stdout output as a string
     */
    public static String execute(String[] args, String workingDir) {
        if (args == null) {
            return "";
        }

        String[] cmd = new String[args.length + 1];
        cmd[0] = LARK_CLI;
        System.arraycopy(args, 0, cmd, 1, args.length);

        try {
            ProcessBuilder pb = new ProcessBuilder(cmd);
            if (workingDir != null && !workingDir.isEmpty()) {
                pb.directory(new File(workingDir));
            }
            pb.redirectErrorStream(true);

            Process process = pb.start();

            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
            StringBuilder output = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                if (output.length() > 0) {
                    output.append('\n');
                }
                output.append(line);
            }
            reader.close();

            process.waitFor();
            return output.toString();

        } catch (IOException e) {
            return "Error: " + e.getMessage();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return "Error: Process interrupted";
        }
    }
}

File: android/src/com/tcc/api/LarkClient.kt
package com.tcc.api

import android.content.Context
import com.tcc.TermuxBootstrap
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader

// lark-cli 命令行包装器
object LarkClient {

    fun hasNode(context: Context): Boolean {
        val prefix = TermuxBootstrap.getPrefixDir(context)
        return File(prefix, "bin/node").isFile
    }

    fun isInstalled(context: Context): Boolean {
        val prefix = TermuxBootstrap.getPrefixDir(context)
        return File(prefix, "bin/lark-cli").isFile
    }

    // 检查 lark-cli 是否可用
    fun isAvailable(context: Context): Boolean {
        if (!hasNode(context) || !isInstalled(context)) return false
        return !execute(context, "--help").startsWith("Error")
    }

    // 执行 lark-cli 命令
    fun execute(context: Context, vararg args: String): String {
        return try {
            val prefix = TermuxBootstrap.getPrefixDir(context)
            val nodePath = File(prefix, "bin/node")
            val cliPath = File(prefix, "bin/lark-cli")
            if (!nodePath.isFile) {
                return "Error: Node.js 未安装，请先安装/修复 Claude 环境"
            }
            if (!cliPath.isFile) {
                return "Error: lark-cli 未安装，请先安装/修复 Claude 环境"
            }
            val pb = TermuxBootstrap.execInTermux(
                context,
                nodePath.absolutePath,
                cliPath.canonicalPath,
                *args
            )

            // Read stdout
            val stdoutReader = BufferedReader(InputStreamReader(pb.inputStream))
            val stdout = stdoutReader.readText()
            stdoutReader.close()

            // Read stderr (for diagnostics)
            val stderrReader = BufferedReader(InputStreamReader(pb.errorStream))
            val stderr = stderrReader.readText()
            stderrReader.close()

            val exitCode = pb.waitFor()

            if (exitCode == 0) {
                stdout.trim()
            } else {
                val errorMsg = stderr.trim().ifEmpty { stdout.trim() }
                "Error (exit $exitCode): $errorMsg"
            }
        } catch (e: Exception) {
            "Error: ${e.message ?: "Unknown error"}"
        }
    }

    // 检查 lark-cli 认证状态
    fun authStatus(context: Context): String {
        return try {
            execute(context, "auth", "status")
        } catch (e: Exception) {
            "Error: ${e.message ?: "Unknown error"}"
        }
    }
}
